### ISSUE TITLE
[New Rule]: no-inner-callback

### ISSUE THREAD

**Issue by anikethsaha** [eslint/eslint#13074]
**Please describe what the rule should do:**
callbacks should always be called at the last of a block. So whenever a callback is used in between a body, it should report.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[X] Other (please specify: ) potential unreachable code

Though this can be added in `no-unreachable` as well, But I think that rule is especially for program control flow statements

**Provide 2-3 code examples that this rule will warn about:**

`Incorrect`

```js
function(cb){
   var a = 'hello';
   cb() // error
   console.log('a', a)
}
```
`Correct`

```js
function(callback){
  var a  = '10'
  callback(a)
}
```

```js
function(cb){
   if(foo) {
      var bar = foo
      cb(bar)
   }
   cb()
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
As callbacks are very common. I think it would be better to add in the core

**Are you willing to submit a pull request to implement this rule?**
Yes 👍

**Comment by kaicataldo** [comment]
Thanks for the proposal!

Can you explain what problem you'd like to solve with this proposal? I'm not sure I understand why callbacks should always be the last statement in a function body. I personally don't think the incorrect example above is particularly error prone or wrong. For a real world example, [here](https://github.com/babel/babel/blob/master/packages/babel-parser/src/parser/statement.js#L890-L893) is some code in Babel that would fail this rule (full disclosure: I wrote this code).

I have some additional questions about how this would work:

1. 

```js
function(cb){
   if(foo) {
      var bar = foo
      cb(bar)
   }
   cb()
}
```

This code would call the callback twice if `foo` is truthy. Can you explain what the rule would actually be checking for? I'm not sure I understand why this is correct but the incorrect example isn't.

---
2. How would this rule handle something like the following?

```js
function example(cb) {
  return thenableFn().then(val => {
    cb(val);
    console.log("success");
  }, err => {
    cb(err);
    console.log("failure");
  });
}
```

```js
async function example(cb) {
  try {
    const val = thenableFn();
    cb(val);
    console.log("success");
  } catch (err) {
    cb(err);
    console.log("failure")
  }
}
```

**Comment by anikethsaha** [comment]
This can be mainly for the **suggestion** category *(I changed in the issuing body)*.
it should be recommended to place the callbacks at the end of the block. 

```js
function(cb){
   if(foo) {
      var bar = foo
      cb(bar)
   }
   cb()
}
```

according to this rule, it will pass cause its not in the scope of this rule to mention that there can be possibilities of cb to get called twice




> Thanks for the proposal!
> 
> Can you explain what problem you'd like to solve with this proposal? I'm not sure I understand why callbacks should always be the last statement in a function body. I personally don't think the incorrect example above is particularly error prone or wrong. For a real world example, [here](https://github.com/babel/babel/blob/master/packages/babel-parser/src/parser/statement.js#L890-L893) is some code in Babel that would fail this rule (full disclosure: I wrote this code).
> 
> I have some additional questions about how this would work:
> 
> 
> ```js
> function(cb){
>    if(foo) {
>       var bar = foo
>       cb(bar)
>    }
>    cb()
> }
> ```
> 
> This code would call the callback twice if `foo` is truthy. Can you explain what the rule would actually be checking for? I'm not sure I understand why this is correct but the incorrect example isn't.
> 
> 1. How would this rule handle something like the following?
> 
> ```js
> function example(cb) {
>   return thenableFn().then(val => {
>     cb(val);
>     console.log("success");
>   }, err => {
>     cb(err);
>     console.log("failure");
>   });
> }
> ```
> 
> ```js
> async function example(cb) {
>   try {
>     const val = await thenableFn();
>     cb(val);
>     console.log("success");
>   } catch (err) {
>     cb(err);
>     console.log("failure")
>   }
> }
> ```

It would result in false negatives. 

Need to rethink this if possible ! closing for now 👍
